### PR TITLE
Changes in regards to issue#3330

### PR DESCRIPTION
**Pull request by ItIsOHM**
## Changes

Moved the `ApplyForcedChanges` method call from `dot/digest/block_import.go` to `dot/core/service.go`

## Issues

- https://github.com/ChainSafe/gossamer/issues/3330

## Primary Reviewer

@timwu20 @EclesioMeloJunior

### CODE REVIEW

**Comment by ItIsOHM** [comment]
Hey @EclesioMeloJunior, I can see that some tests have failed. Is there a formatting issue?

**Comment by jimjbrettj** [comment]
Hey, @ItIsOHM thanks for the contribution! I think there are some build issues, it seems that there are missing mock calls so I think you will need to update the tests to have the correct mocks usage with these changes.

For additional context, when we use mocks in our tests it will fail if there is an unexpected call on a mocked interface or if an expected call does not occur. I think that in the case where the function was removed there are now missing mock calls, and in the core package there are now unexpected ones.

Hope this makes sense and I am happy to help out or answer more questions if you have any!

**Comment by ItIsOHM** [comment]
Oh i see, thanks for the explanation @jimjbrettj! So from what I understand is i have to update all the test files in `tests/`? Or only some specific test files since there are a bunch of them present :D

**Comment by jimjbrettj** [comment]
> Oh i see, thanks for the explanation @jimjbrettj! So from what I understand is i have to update all the test files in `tests/`? Or only some specific test files since there are a bunch of them present :D

So actually you do not need to worry about the tests in the tests dir. The way our code is structured is that unit and integration tests are located in the same package as the logic itself. For example, in this case you made a change to dot/core/service.go, so the associated tests for this file are service_test.go and service_integration_test.go.

You can see this example of mocking from our service_test.go file https://github.com/ChainSafe/gossamer/blob/457fe39f7a12cb9eb04de5b6a315d38825247d68/dot/core/service_test.go#L336. Here you can see how we define what interfaces we are mocking and the assert the mocked calls. This is the logic you will need to change in the tests for the functions you modified.

What I would do is run the tests for the packages you change (i.e. go test ./dot/core/ -v) and then see where it fails, then that is the test you need to modify.

Hope this provides some clarity and am happy to answer any more questions!

**Comment by EclesioMeloJunior** [comment]
@ItIsOHM thanks for the PR! Some thinks you should change:

The `Service` struct in `dot/core/service.go` does not contain a reference to `grandpaState`, so you need to inject the `GrandpaState` (dot/state/grandpa.go) in the `core.Service`, you can inject through the `core.NewService` function. Similar to what is done in the `Handler` struct (dot/digest/digest.go).

1. move the GrandpaState interface from `dot/digest/interfaces.go` and place it in `dot/core/interfaces.go`
2. At `dot/core/service.go` include a new property in the `Service` called `grandpaState` and in `Config` called `GrandpaState` which in both cases holds the implementation of the `GrandpaState` interface we moved earlier
```go
type Service struct {
	...
        grandpaState GrandpaState
}

// Config holds the configuration for the core Service.
type Config struct {
	...
        GrandpaState GrandpaState // add this new property
}
```
3. At `dot/core/service.go` in the `NewService` you pass the injected reference, that comes in the Config struct to the `Service` instance being created
```go
// NewService returns a new core service that connects the runtime, BABE
// session, and network service.
func NewService(cfg *Config) (*Service, error) {
        ...
	srv := &Service{
		...
                grandpaState: cfg.GrandpaState
	}

	return srv, nil
}
```

4. At this point the `core.Service` expects a `GrandpaState` instance to be injected, and you will do that in the file `dot/services.go` in the method `createCoreService`

```go
func (nodeBuilder) createCoreService(config *cfg.Config, ks *keystore.GlobalKeystore,
	st *state.Service, net *network.Service, dh *digest.Handler) (
	*core.Service, error) {
        ...
	// set core configuration
	coreConfig := &core.Config{
		...
                GrandpaState: st.Grandpa // here you are injecting the grandpa state instance in the core service
	}

	// create new core service
	coreSrvc, err := core.NewService(coreConfig)
	if err != nil {
		logger.Errorf("failed to create core service: %s", err)
		return nil, err
	}

	return coreSrvc, nil
}
```

5. Now you completely move the method call from `digest` service to `core` service, but there is some clean up will need to execute, 1) The `BlockImportHandler` (dot/digest/block_import.go) don't need the GrandpaState anymore, so you can remove it from there:

```go
type BlockImportHandler struct {
	epochState   EpochState
	grandpaState GrandpaState // <- remove
}

func NewBlockImportHandler(epochState EpochState, grandpaState GrandpaState  /* <- remove */ ) *BlockImportHandler {
	return &BlockImportHandler{
		epochState:   epochState,
		grandpaState: grandpaState,  // <- remove
	}
}
```
 
6. Finally you can execute `make build` and `./bin/gossamer --chain westend-dev --alice` and gossamer should start producing blocks successfully.

**Comment by ItIsOHM** [comment]
Hey there! Thanks a ton for the detailed guide, really helped me to understand the logics behind this :D 

I made the changes as @EclesioMeloJunior suggested but couldn't build gossamer due to some issues like `undefined: cWasmerImportObjectT`, I updated the wasmer-go repo but still couldn't build it.

Further, I uncommented this line from `dot\core\mocks_generate_test.go` to run and generate the mock tests for GrandpaState interface: `go:generate mockgen -destination=mocks_test.go -package $GOPACKAGE . BlockState,StorageState,TransactionState,Network,CodeSubstitutedState,Telemetry,BlockImportDigestHandler,GrandpaState`

I hope this time it should pass the tests but I apologize if I made the matter worse.

**Comment by EclesioMeloJunior** [comment]
@ItIsOHM if you're using a apple M1 processor then this could help you https://github.com/ChainSafe/gossamer#troubleshooting-for-apple-silicon-users

**Comment by ItIsOHM** [comment]
> @ItIsOHM if you're using a apple M1 processor then this could help you https://github.com/ChainSafe/gossamer#troubleshooting-for-apple-silicon-users

Hehe I actually use a Windows 11 system, but still I added the GOARCH path and value = amd64 to my system and env variables, but that didn't help too.

**Comment by EclesioMeloJunior** [comment]
> > @ItIsOHM if you're using a apple M1 processor then this could help you https://github.com/ChainSafe/gossamer#troubleshooting-for-apple-silicon-users
> 
> Hehe I actually use a Windows 11 system, but still I added the GOARCH path and value = amd64 to my system and env variables, but that didn't help too.

@ItIsOHM nice! what if you set GOARCH to `""` and try to build?

**Comment by EclesioMeloJunior** [comment]
@ItIsOHM regards the `mocks_generate_test.go` file you should keep the line commented, that is how the golang generate workers

**Comment by ItIsOHM** [comment]
> @ItIsOHM regards the `mocks_generate_test.go` file you should keep the line commented, that is how the golang generate works

I see! I think the latest commit should have added the comments back and also added go.mod and go.sum back?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/digest/block_import.go [hunk 1]
```diff
@@ -69,10 +60,12 @@ func (h *BlockImportHandler) handleConsensusDigest(d *types.ConsensusDigest, hea
...
-		err = h.grandpaState.HandleGRANDPADigest(header, data)
-		if err != nil {
-			return fmt.Errorf("handling grandpa digest: %w", err)
-		}
+		// Shouldn't be needed since the grandpaState is no longer in use here.
+		// err = h.grandpaState.HandleGRANDPADigest(header, data)
+		// if err != nil {
+		// 	return fmt.Errorf("handling grandpa digest: %w", err)
+		// }
+
```
Comment (EclesioMeloJunior) [inline dot/digest/block_import.go:68]: you can remove this comment

File: dot/core/service.go [hunk 2]
```diff
@@ -92,6 +95,7 @@ func NewService(cfg *Config) (*Service, error) {
 		codeSubstitute:       cfg.CodeSubstitutes,
 		codeSubstitutedState: cfg.CodeSubstitutedState,
 		onBlockImport:        cfg.OnBlockImport,
+		grandpaState: 		  cfg.grandpaState,
```
Comment (EclesioMeloJunior) [inline dot/core/service.go:98]:
```suggestion
		grandpaState: 		  cfg.GrandpaState,
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]
Please, execute `go fmt`, it should included the missing imports and will adust the identation as well

File: dot/digest/block_import.go (hunk 1 above)
Comment (ItIsOHM) [reply dot/digest/block_import.go:68]: Do you mean I should remove this entire block, or just the comment in line 63 and uncomment the rest?

**Comment by ItIsOHM** [comment]
Hey @EclesioMeloJunior does this commit seem to fix the formatting?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

**Comment by EclesioMeloJunior** [comment]
@ItIsOHM you need to run `go generate -run mockgen ./dot/core/...` to update the mocks as you've added a new interface in the `core` package

**Comment by ItIsOHM** [comment]
> @ItIsOHM you need to run `go generate -run mockgen ./dot/core/...` to update the mocks as you've added a new interface in the `core` package

@EclesioMeloJunior i'm not able to generate the new mocks. When i run `go generate -run mockgen ./dot/core/...` I get the following error: 
image

After I add `protobuf`, some new dependencies are added in the `go.mod` and `go.sum` file. After running `go generate -run mockgen ./dot/core/...` again, I again receive an error regarding some more dependencies that I need to install. After installing them, I once again get a bunch of errors:
image

I checked online regarding the `go.mod file not found...` error but those fixes can't seem to make the errors go away.

Can you please look into this and help me out? I'm terribly sorry for all this fuss I'm causing on such an issue.

**Comment by dimartiro** [comment]
@ItIsOHM There is an import missing in `dot/core/interfaces.go`, you should add `"github.com/ChainSafe/gossamer/pkg/scale"` in the imports list.

I added it and the `go generate -run mockgen ./dot/core/...` worked for me

**Comment by ItIsOHM** [comment]
Oh alright! I'll make the changes and update you guys :D
Thanks @dimartiro!

**Comment by EclesioMeloJunior** [comment]
@ItIsOHM I made a PR to you fork fixing the build issue and the testing, could you take a look and merge it https://github.com/ItIsOHM/gossamer/pull/1

**Comment by ItIsOHM** [comment]
> @ItIsOHM I made a PR to you fork fixing the build issue and the testing, could you take a look and merge it [ItIsOHM#1](https://github.com/ItIsOHM/gossamer/pull/1)

Just did! Also I added the suggestion by @dimartiro in my interfaces.go, and yet I received the same error messages involving wasmerio. Any idea why? Is it a platform issue?

**Comment by dimartiro** [comment]
@ItIsOHM Are you planning to continue with this PR? We changed wasmerio to wazero so I think your issue should be solved, can you check it?

**Comment by ItIsOHM** [comment]
> @ItIsOHM Are you planning to continue with this PR? We changed wasmerio to wazero so I think your issue should be solved, can you check it?

No i'm sorry @dimartiro, I can't continue with this PR.

**Comment by dimartiro** [comment]
> > @ItIsOHM Are you planning to continue with this PR? We changed wasmerio to wazero so I think your issue should be solved, can you check it?
> 
> No i'm sorry @dimartiro, I can't continue with this PR.

So sad to read that @ItIsOHM but thank you for your contribution anyway 🙌

**Comment by jimjbrettj** [comment]
Hey @ItIsOHM I have just opened a PR to finish up this work (#3464) and perhaps it can provide some clarification to any issues you may have faced. Appreciate your work and if you have any follow-up questions on this or my PR, feel free to ask!

**Comment by jimjbrettj** [comment]
Replaced by #3464

**Comment by ItIsOHM** [comment]
> Hey @ItIsOHM I have just opened a PR to finish up this work (#3464) and perhaps it can provide some clarification to any issues you may have faced. Appreciate your work and if you have any follow-up questions on this or my PR, feel free to ask!

Thanks so much, @jimjbrettj! Just checked it out, and it made me realize what all I was doing wrong and how it could be made better in the future! Also, I'm once again grateful to all the moderators and reviewers here for putting up with my naive questions :D

**Pull request closed without merging**